iree/hal/allocator_caching.c

```diff
@@ -0,0 +1,220 @@
+
+static bool iree_hal_caching_allocator_buffer_available(iree_host_size_t requested_size) {
+  iree_hal_caching_allocator_cache
+    *iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_list;
+  while(iree_hal_caching_allocator_cache_ptr != NULL) {
+    if(iree_hal_caching_allocator_cache_ptr->buffer->allocation_size >= requested_size) {
+      return true;
+    }
+    iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_ptr->next;
+  }
+  return false;
+}
+
+static void iree_hal_caching_allocator_remove_buffer_from_cache(int buffer_index_to_remove) {
+  iree_hal_caching_allocator_cache *buffer_to_be_removed_from_cache,
+    *iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_list;
+  if(buffer_index_to_remove == 0) {
+    buffer_to_be_removed_from_cache = iree_hal_caching_allocator_cache_ptr;
+    if(buffer_to_be_removed_from_cache->next == NULL) {
+      iree_hal_caching_allocator_cache_list = NULL;
+    } else {
+      iree_hal_caching_allocator_cache_list = iree_hal_caching_allocator_cache_list->next;
+    }
+  } else {
+    int current_buffer_index = 0;
+    while(current_buffer_index < buffer_index_to_remove-1) {
+      iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_ptr->next;
+      current_buffer_index++;
+    }
+    buffer_to_be_removed_from_cache = iree_hal_caching_allocator_cache_ptr->next;
+
+    if(buffer_to_be_removed_from_cache->next == NULL) {
+      iree_hal_caching_allocator_cache_ptr->next = NULL;
+    } else {
+      iree_hal_caching_allocator_cache_ptr->next = buffer_to_be_removed_from_cache->next;
+    }
+  }
+  free(buffer_to_be_removed_from_cache);
+  return;
+}
+
+static iree_status_t iree_hal_caching_allocator_allocate_from_cache(
+    iree_host_size_t requested_size, iree_hal_buffer_t** out_buffer) {
+  size_t buffer_index_in_cache = 0;
+  iree_hal_caching_allocator_cache* 
+    iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_list;
+  while(iree_hal_caching_allocator_cache_ptr != NULL) {
+    if(iree_hal_caching_allocator_cache_ptr->buffer->allocation_size >= requested_size) {
+      break;
+    }
+    iree_hal_caching_allocator_cache_ptr = iree_hal_caching_allocator_cache_ptr->next;
+    buffer_index_in_cache++;
+  }
+  iree_hal_buffer_t *buffer_to_allocate = iree_hal_caching_allocator_cache_ptr->buffer;
+  
+  *out_buffer = buffer_to_allocate;
+  iree_hal_caching_allocator_remove_buffer_from_cache(buffer_index_in_cache);
+
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT iree_status_t iree_hal_allocator_add_buffer_to_cache(iree_hal_buffer_t* buffer) {
+  struct iree_hal_caching_allocator_cache* cache_buffer_node = 
+    (iree_hal_caching_allocator_cache*) malloc(sizeof(iree_hal_caching_allocator_cache));
+  cache_buffer_node->buffer = buffer;
+  cache_buffer_node->next = NULL;
+
+  if(iree_hal_caching_allocator_cache_list == NULL) {
```

Comment:
Is there a reason why you can't just add the buffer to the head of the list? That way, you just modify the head pointer and don't need to traverse the whole list.